Get rid of System from ThreadedElementLoop

fdkong opened the pull request:
`ThreadedElementLoop` does not necessarily depend on
`System`.

Refs #7398 #7654

fdkong committed:
Get rid of System from ThreadedElementLoop

ThreadedElementLoop does not necessarily depend on
System.

Refs #7398 #7654

andrsd commented:
In general, this is ok, but can be even improved in a following way. When we pass in `FEProblem` and `NonlinearSystem` as parameters in the c-tor, we actually need just the `FEProblem`. The instance of `NonlinearSystem` can be obtained in the initializer list as `_nl(_fe_problem.getNonlinearSystem()),`.

Similarly for `AuxiliarySystem`...

friedmud commented:
Ummm... the whole purpose of `ThreadedElementLoopBase` is to be a "threaded element loop" without a System.  Just use that instead if you need to do that.

fdkong commented:
> In general, this is ok, but can be even improved in a following way. When we pass in FEProblem and NonlinearSystem as parameters in the c-tor, 

For which classes? For any class which has a constructor like `Class(FEProblem &, NonlinearSystem/AuxiliarySystem &)` ?

>we actually need just the FEProblem. The instance of NonlinearSystem can be obtained in the initializer list as _nl(_fe_problem.getNonlinearSystem()),.
> 
> Similarly for AuxiliarySystem...

andrsd commented:
> For any class which has a constructor like Class(FEProblem &, NonlinearSystem/AuxiliarySystem &) ?

Yes.

fdkong commented:
> Ummm... the whole purpose of ThreadedElementLoopBase is to be a "threaded element loop" without a System. Just use that instead if you need to do that.

I don't  see any good reason to let `ThreadedElementLoopBase` depend on `System` either.

The dependency also causes some issues. Please see comments  we have at #7654.

fdkong commented:
> For any class which has a constructor like Class(FEProblem &, NonlinearSystem/AuxiliarySystem &) ?
> 
> Yes.

There a lot of classes like this. Let me improve this in another PR.

andrsd commented:
> Let me improve this in another PR.

Why another PR? Another patch within this PR is going to be just fine...

fdkong commented:
> Why another PR? Another patch within this PR is going to be just fine...

I am fine with this. I just won't give reviewers have hard time

friedmud commented:
`ThreadedElementLoopBase` does _not_ depend on `System`.  That is the point of that class.

andrsd commented:
[This](https://github.com/idaholab/moose/pull/7654#issuecomment-248159650) is the motivation behind this change. The point is that not even the child classes need `System` passed in via c-tor...

friedmud commented:
Ok - so are you going to remove `ThreadedElementLoopBase`?

andrsd commented:
I did not have that in mind, but it looks like it should be possible...

fdkong commented:
> Ok - so are you going to remove ThreadedElementLoopBase?

Possibly not. `ThreadedElementLoopBase` depends on `MooseMesh` only but `ThreadedElementLoop` also depends on `FEProblem` for setting an exception something like that.

andrsd commented:
I think that we never loop over the displaced version of the mesh, so having `FEProblem` gives us the right `MooseMesh` that the `ThreadedElementLoopBase` needs right now. Thus, it should be possible to merge the `ThreadedElementLoopBase` into `ThreadedElementLoop`...

fdkong committed:
Remove NonlinearSystem/AxiliarySystem from c-tors

Remove NonlinearSystem/AxiliarySystem from constructors as long as FEProblem is also an input parameter.
Nonlinearsystem and AxiliarSystem can be retreived from FEProblem.

Refs  #7398 #7654

fdkong commented:
> I think that we never loop over the displaced version of the mesh, so having FEProblem gives us the right MooseMesh that the ThreadedElementLoopBase needs right now. Thus, it should be possible to merge the ThreadedElementLoopBase into ThreadedElementLoop...

Do you want this in the same PR too, @andrsd ?

andrsd commented:
> Do you want this in the same PR too, @andrsd ?

Yes.

fdkong commented:
@YaqiWang's comments disappeared after I did something like:

`git commit -a --amend `

`git push -f origin get_rid_of_system_from_ThreadedElementLoop`

fdkong commented:
@andrsd, `CacheChangedListsThread` inherits from `ThreadedElementLoopBase` directly.  `CacheChangedListsThread` is used in `MooseMesh` where we do not have access to `FEProblem` that is supposed as an input parameter for `ThreadedElementLoop`. 

If we merge `ThreadedElementLoopBase` into `ThreadedElementLoop`, then `CacheChangedListsThread` has to derive from `ThreadedElementLoop` instead of `ThreadedElementLoopBase`. In `MooseMesh`, we have no way to pass `FEProblem` to  `CacheChangedListsThread`.

andrsd commented:
Ok

fdkong commented:
> Ok

Is this PR ready to go? Or you need more time to review again?

YaqiWang commented:
To me it is ok. But framework developers may have more comments.

fdkong commented:
> To me it is ok. But framework developers may have more comments.

@idaholab/moose-team, any comments?

permcody reviewed (commented):
This looks good to me 👍

permcody commented:
@andrsd Please merge since you've been guiding @fdkong on this change.

andrsd commented:
I am on it...

andrsd commented on framework/include/base/ComputeFullJacobianThread.h:
```diff
@@ -39,6 +39,9 @@ class ComputeFullJacobianThread : public ComputeJacobianThread
   virtual void computeInternalFaceJacobian(const Elem * neighbor) override;
   virtual void computeInternalInterFaceJacobian(BoundaryID bnd_id) override;
 
+  // Reference to NonlinearSystem
```
Remove this line, this is cpt. Obvious...

fdkong replied:
Done
